scipy/spatial/transform/_rotation.pyx

```diff
@@ -311,23 +311,134 @@ cdef double[:, :] _compute_euler_from_quat(
         # compute first and third angles, according to case
         half_sum = atan2(b, a)
         half_diff = atan2(d, c)
-        
+
         if case == 0:  # no singularities
             _angles[angle_first] = half_sum - half_diff
             _angles[angle_third] = half_sum + half_diff
-        
+
         else:  # any degenerate case
             _angles[2] = 0
             if case == 1:
                 _angles[0] = 2 * half_sum
             else:
                 _angles[0] = 2 * half_diff * (-1 if extrinsic else 1)
-                
+
         # for Tait-Bryan angles
         if not symmetric:
             _angles[angle_third] *= sign
             _angles[1] -= pi / 2
-            
+
+        for idx in range(3):
+            if _angles[idx] < -pi:
+                _angles[idx] += 2 * pi
+            elif _angles[idx] > pi:
+                _angles[idx] -= 2 * pi
+
+        if case != 0:
+            warnings.warn("Gimbal lock detected. Setting third angle to zero "
+                          "since it is not possible to uniquely determine "
+                          "all angles.")
+
+    return angles
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+cdef double[:, :] _compute_davenport_from_quat(
+    np.ndarray[double, ndim=2] quat, np.ndarray[double, ndim=1] n1,
+        np.ndarray[double, ndim=1] n2, np.ndarray[double, ndim=1] n3,
+        bint extrinsic=False, bint correct_set=False
+):
+    # The algorithm assumes extrinsic frame transformations. The algorithm
+    # in the paper is formulated for rotation quaternions, which are stored
+    # directly by Rotation.
+    # Adapt the algorithm for our case by reversing both axis sequence and
+    # angles for intrinsic rotations when needed
+
+    # intrinsic/extrinsic conversion helpers
+    cdef int angle_first, angle_third
+    if extrinsic:
+        angle_first = 0
+        angle_third = 2
+    else:
+        n1, n3 = n3, n1
+        angle_first = 2
+        angle_third = 0
+
+    cdef double[:] n_cross = _cross3(n1, n2)
+    cdef double lamb = atan2(_dot3(n3, n_cross), _dot3(n3, n1))
+
+    if correct_set and lamb < 0:
+        # alternative set of angles compatible with as_euler implementation
+        n2 = -n2
+        lamb = -lamb
+        n_cross[0] = -n_cross[0]
+        n_cross[1] = -n_cross[1]
+        n_cross[2] = -n_cross[2]
+    else:
+        correct_set = False
+
+    cdef double[:] quat_lamb = np.array([
+            sin(lamb / 2) * n2[0],
```

Comment:
Might consider precomputing sin and cos.

Reply:
`quat_lamb` is a constant quaternion, so it is only computed once. Do you think it is worth it to add an extra line for pre-computing `sin(lamb / 2)` in this case? It would save two operations overall, but I assumed this was probably done by the compiler...

Reply:
Yes, most likely, you are right.

This Cython low level syntax is kind of unpleasant though.

Reply:
Should I leave it like this then?

Reply:
Yes, no problem.